### ISSUE TITLE
skydns container in reboot loop on minion

### ISSUE THREAD

**Issue by tleyden** [kubernetes/kubernetes#16243]
## Steps to reproduce

* Install Kubernetes via `KUBERNETES_PROVIDER=vagrant curl -sS https://get.k8s.io | bash`
* ssh into kubernetes minion
* run `docker ps` a few times
* See that the container is [only up a few seconds](https://gist.github.com/tleyden/807aee9c4daa64973425#file-gistfile1-txt-L3) and then [later restarts](https://gist.github.com/tleyden/807aee9c4daa64973425#file-gistfile1-txt-L12)

## Setup

* OSX Yosemite
* VirtualBox 5.0.8
* Vagrant 1.7.4
* Kubernetes version: `Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.1", GitCommit:"6a5c06e3d1eb27a6310a09270e4a5fb1afa93e74", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}`

Seems like it might be a duplicate of https://github.com/kubernetes/kubernetes/issues/8042

**Comment by thockin** [comment]
Can you try disabling the kube-dns pod's healthcheck for skydns?  Assuming
that stops t from crashing, try to manually do hat health check (kubectl
exec ... Nslookup ...).

**Comment by ArtfulCoder** [comment]
It might be helpful to see kube2sky logs if you could check that as well..

**Comment by tleyden** [comment]
@ArtfulCoder:

Here are the kube2sky container logs:

```
[root@kubernetes-minion-1 vagrant]# docker logs e1a87e875ccf
2015/10/25 19:31:44 skydns: falling back to default configuration, could not read from etcd: 100: Key not found (/skydns) [2]
2015/10/25 19:31:44 skydns: ready for queries on cluster.local. for tcp://0.0.0.0:53 [rcache 0]
2015/10/25 19:31:44 skydns: ready for queries on cluster.local. for udp://0.0.0.0:53 [rcache 0]
[root@kubernetes-minion-1 vagrant]# docker logs 048610bdd04f
I1024 23:20:34.834580       1 kube2sky.go:389] Etcd server found: http://127.0.0.1:4001
I1024 23:20:35.837332       1 kube2sky.go:455] Using https://10.247.0.1:443 for kubernetes master
I1024 23:20:35.837347       1 kube2sky.go:456] Using kubernetes API
E1024 23:20:38.848758       1 reflector.go:136] Failed to list *api.Service: Get https://10.247.0.1:443/api/v1/services: dial tcp 10.247.0.1:443: no route to host
E1024 23:20:38.850105       1 reflector.go:136] Failed to list *api.Endpoints: Get https://10.247.0.1:443/api/v1/endpoints: dial tcp 10.247.0.1:443: no route to host
```

(The last two lines are repeated hundreds of times)

Also here are the skydns container logs:

```
[root@kubernetes-minion-1 vagrant]# docker logs e1a87e875ccf
2015/10/25 19:31:44 skydns: falling back to default configuration, could not read from etcd: 100: Key not found (/skydns) [2]
2015/10/25 19:31:44 skydns: ready for queries on cluster.local. for tcp://0.0.0.0:53 [rcache 0]
2015/10/25 19:31:44 skydns: ready for queries on cluster.local. for udp://0.0.0.0:53 [rcache 0]
```

**Comment by tleyden** [comment]
@thockin 

> Can you try disabling the kube-dns pod's healthcheck for skydns? 

Sure, how do I do that?

>  try to manually do hat health check (kubectl exec ... Nslookup ...).

Can you give me the exact command?

**Comment by ArtfulCoder** [comment]
The issue seems to be that kube2sky can't reach API server.
Our health check is slightly simplistic and assumes that skydns container is the problem and restarts that.

**Comment by tleyden** [comment]
@ArtfulCoder I just noticed in the logs it seems to be connecting to a different IP than what `kubectl cluster-info` returns.  Could that be an issue?

```
$ kubectl cluster-info
Kubernetes master is running at https://10.245.1.2
KubeDNS is running at https://10.245.1.2/api/v1/proxy/namespaces/kube-system/services/kube-dns
KubeUI is running at https://10.245.1.2/api/v1/proxy/namespaces/kube-system/services/kube-ui
```

**Comment by ArtfulCoder** [comment]
You should check the service ip of Kubernetes.
Can you also run sudo iptables-save on the minion.

**Comment by tleyden** [comment]
```
$ kubectl get services
NAME         LABELS                                    SELECTOR   IP(S)        PORT(S)
kubernetes   component=apiserver,provider=kubernetes   <none>     10.247.0.1   443/TCP
```

But if I ssh into the minion, I'm not able to ping `10.247.0.1`, whereas I am able to ping `10.245.1.2` (the ip listed in cluster-info)

**Comment by tleyden** [comment]
Here's the output from [iptables-save](https://gist.github.com/tleyden/61576f574a26686618cd)

**Comment by ArtfulCoder** [comment]
so you have service-ip entry in iptables.
Can you run kubectl get ep and see if you see the kubernetes endpoint ?

**Comment by tleyden** [comment]
```
$ kubectl get ep
NAME         ENDPOINTS
kubernetes   10.245.1.2:443
```

**Comment by tleyden** [comment]
From the minion, I'm able to telnet to both `10.247.0.1 443` and `10.245.1.2 443`, so I don't get why it wouldn't be reachable from the container.

**Comment by ArtfulCoder** [comment]
instead of the minion,can you telnet from the healthcheck/kube2sky/etcd pod on the minion ? (whichever of those 3 have telnet on it)

**Comment by thockin** [comment]
FWIW service IPs don't ping.  But Abhi has you on the right path.

**Comment by liqinghua188** [comment]
I installed kubernetes On Bare-metal Ubuntu Nodes， following [http://kubernetes.io/v1.0/docs/getting-started-guides/ubuntu.html](http://kubernetes.io/v1.0/docs/getting-started-guides/ubuntu.html)

My kubectl version is :

	kubernetes-master@kubernetes-master:/srv/kubernetes$ kubectl version
	Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
	Server Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}


I met the similar problem，here are the logs on kube2sky	


	E1026 02:22:02.121862       1 reflector.go:136] Failed to list *api.Service: Get https://11.1.1.1:443/api/v1/services: read tcp 11.1.1.1:443: connection reset by peer
	E1026 02:22:02.249474       1 reflector.go:136] Failed to list *api.Endpoints: Get https://11.1.1.1:443/api/v1/endpoints: read tcp 11.1.1.1:443: connection reset by peer
	E1026 02:22:03.123928       1 reflector.go:136] Failed to list *api.Service: Get https://11.1.1.1:443/api/v1/services: read tcp 11.1.1.1:443: connection reset by peer
	E1026 02:22:03.253032       1 reflector.go:136] Failed to list *api.Endpoints: Get https://11.1.1.1:443/api/v1/endpoints: read tcp 11.1.1.1:443: connection reset by peer


11.1.1.1 is the ip of kubernetes service, 

	kubernetes-master@kubernetes-master:/srv/kubernetes$ kubectl get svc
	NAME                      LABELS                                    SELECTOR              IP(S)        PORT(S)
	kubernetes                component=apiserver,provider=kubernetes   <none>                11.1.1.1     443/TCP

I'm able to telnet 11.1.1.1:443, howerver 

	kubernetes-node2@kubernetes-node2:~$ curl https://11.1.1.1:443/api/v1/endpoints
	curl: (35) Unknown SSL protocol error in connection to 11.1.1.1:443 

so， i debt whether becuase I don‘t have the apiServer's authetication?  How I configure the ca file on kube2sky container or on the minon node？

**Comment by GuoHui89757** [comment]
@liqinghua188 
      - name: kube2sky
        image: gcr.io/google_containers/kube2sky:1.11
        resources:
          limits:
            cpu: 100m
            memory: 50Mi
        args:
        # command = "/kube2sky"
        - -domain={{ pillar['dns_domain'] }}
        - -kube_master_url=http://k8s-master-ip:8080
        - -etcd-server=http://127.0.0.1:4001
        - -logtostderr
        - -v=4
Maybe you can try this.

**Comment by liqinghua188** [comment]
@GuoHui89757 
Great! it works for me. thank you

**Comment by tleyden** [comment]
@GuoHui89757 what is the full path to that file?  Also, can you post the diff of what you changed?

**Comment by tleyden** [comment]
@ArtfulCoder 

> instead of the minion,can you telnet from the healthcheck/kube2sky/etcd pod on the minion ? (whichever of those 3 have telnet on it)

Sorry, not exactly sure what you mean.  Do you want me to `docker exec` into those three containers and then try to run telnet assuming it is installed?

**Comment by tleyden** [comment]
@ArtfulCoder if that's what you meant, I need some help .. how do I get a bash shell on those containers?

```
[vagrant@kubernetes-minion-1 ~]$ sudo docker exec -ti 048610bdd04f bash
exec: "bash": executable file not found in $PATH
```

**Comment by ArtfulCoder** [comment]
just try sudo docker exec -ti 048610bdd04f telnet ....

also, if you want to get to a shell, and bash is not present, try "sh"

**Comment by tleyden** [comment]
I just tried it on the kube2sky container:

```
[vagrant@kubernetes-minion-1 ~]$ sudo docker exec -ti 048610bdd04f sh
/ # telnet 10.247.0.1 443
telnet: can't connect to remote host (10.247.0.1): No route to host
/ # telnet 10.245.1.2 443
telnet: can't connect to remote host (10.245.1.2): No route to host
```

Also, tried spawning a new container and got the same result:

```
[vagrant@kubernetes-minion-1 ~]$ sudo docker run -ti --rm --entrypoint=/bin/sh gcr.io/google_containers/kube2sky:1.11
/ # telnet 10.247.0.1 443
telnet: can't connect to remote host (10.247.0.1): No route to host
/ # telnet 10.245.1.2 443
telnet: can't connect to remote host (10.245.1.2): No route to host
/ # exit
```

**Comment by GuoHui89757** [comment]
@tleyden 
The file path is kubernetes/cluster/addons/dns/skydns-rc.yaml.in
@@ -48,8 +48,12 @@ spec:
         args:
         # command = "/kube2sky"
         - -domain={{ pillar['dns_domain'] }}
+        - -kube_master_url=http://192.168.129.111:8080
+        - -etcd-server=http://127.0.0.1:4001
+        - -logtostderr
+        - -v=4
"http://192.168.129.111:8080" is my k8s master addr.

**Comment by liqinghua188** [comment]
@GuoHui89757 

i am very sorry to tell you my dns still can not work correctly. the skydns logs :

	kubernetes-master@kubernetes-master:~$ kubectl logs -f kube-dns-v9-mwrc1 -c skydns --namespace=kube-system
	2015/10/14 08:29:10 skydns: falling back to default configuration, could not read from etcd: 100: Key not found (/skydns) [1]
	2015/10/14 08:29:10 skydns: ready for queries on cluster.local. for tcp://0.0.0.0:53 [rcache 0]
	2015/10/14 08:29:10 skydns: ready for queries on cluster.local. for udp://0.0.0.0:53 [rcache 0] 

the kube2sky logs :

	I1014 08:41:24.384633       1 kube2sky.go:94] Removing kubernetes.default.cluster.local. from DNS
	I1014 08:41:24.388828       1 kube2sky.go:94] Removing kubernetes.default.svc.cluster.local. from DNS
	I1014 08:41:24.395088       1 kube2sky.go:246] Setting DNS record: kubernetes.default.cluster.local. -> "{\"host\":\"11.1.1.1\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: kubernetes.default.cluster.local.
	I1014 08:41:24.398010       1 kube2sky.go:246] Setting DNS record: kubernetes.default.svc.cluster.local. -> "{\"host\":\"11.1.1.1\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: 2f15de08.kubernetes.default.svc.cluster.local.
	I1014 08:41:24.400139       1 kube2sky.go:94] Removing kube-dns.kube-system.cluster.local. from DNS
	I1014 08:41:24.402912       1 kube2sky.go:94] Removing kube-dns.kube-system.svc.cluster.local. from DNS
	I1014 08:41:24.405130       1 kube2sky.go:246] Setting DNS record: kube-dns.kube-system.cluster.local. -> "{\"host\":\"11.1.1.10\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: kube-dns.kube-system.cluster.local.
	I1014 08:41:24.408975       1 kube2sky.go:246] Setting DNS record: kube-dns.kube-system.svc.cluster.local. -> "{\"host\":\"11.1.1.10\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: 7dfea3c6.kube-dns.kube-system.svc.cluster.local.
	I1014 08:41:24.414847       1 kube2sky.go:94] Removing kube-ui.kube-system.cluster.local. from DNS
	I1014 08:41:24.417809       1 kube2sky.go:94] Removing kube-ui.kube-system.svc.cluster.local. from DNS
	I1014 08:41:24.419919       1 kube2sky.go:246] Setting DNS record: kube-ui.kube-system.cluster.local. -> "{\"host\":\"11.1.1.151\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: kube-ui.kube-system.cluster.local.
	I1014 08:41:24.422180       1 kube2sky.go:246] Setting DNS record: kube-ui.kube-system.svc.cluster.local. -> "{\"host\":\"11.1.1.151\",\"priority\":10,\"weight\":10,\"ttl\":30,\"targetstrip\":0}", with recordKey: 913d376.kube-ui.kube-system.svc.cluster.local.
	E1014 08:46:52.647498       1 reflector.go:183] watch of *api.Service ended with: EOF
	E1014 08:46:52.665966       1 reflector.go:183] watch of *api.Endpoints ended with: EOF
	E1014 08:46:53.648723       1 reflector.go:136] Failed to list *api.Service: Get http://10.10.103.250:8080/api/v1/services: dial tcp 10.10.103.250:8080: connection refused
	E1014 08:46:53.668124       1 reflector.go:136] Failed to list *api.Endpoints: Get http://10.10.103.250:8080/api/v1/endpoints: dial tcp 10.10.103.250:8080: connection refused

10.10.103.250 is my master ip, and i am able to telnet 10.10.103.250:8080 on another node.

the etcd logs are normal:

	kubernetes-master@kubernetes-master:~$ kubectl logs -f kube-dns-v9-mwrc1 -c etcd --namespace=kube-system
	2015/10/14 08:29:10 etcd: listening for peers on http://localhost:2380
	2015/10/14 08:29:10 etcd: listening for peers on http://localhost:7001
	2015/10/14 08:29:10 etcd: listening for client requests on http://127.0.0.1:2379
	2015/10/14 08:29:10 etcd: listening for client requests on http://127.0.0.1:4001
	2015/10/14 08:29:10 etcdserver: datadir is valid for the 2.0.1 format
	2015/10/14 08:29:10 etcdserver: name = default
	2015/10/14 08:29:10 etcdserver: data dir = /var/etcd/data
	2015/10/14 08:29:10 etcdserver: member dir = /var/etcd/data/member
	2015/10/14 08:29:10 etcdserver: heartbeat = 100ms
	2015/10/14 08:29:10 etcdserver: election = 1000ms
	2015/10/14 08:29:10 etcdserver: snapshot count = 10000
	2015/10/14 08:29:10 etcdserver: advertise client URLs = http://127.0.0.1:2379,http://127.0.0.1:4001
	2015/10/14 08:29:10 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
	2015/10/14 08:29:10 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
	2015/10/14 08:29:10 etcdserver: start member 6a5871dbdd12c17c in cluster f68652439e3f8f2a
	2015/10/14 08:29:10 raft: 6a5871dbdd12c17c became follower at term 0
	2015/10/14 08:29:10 raft: newRaft 6a5871dbdd12c17c [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
	2015/10/14 08:29:10 raft: 6a5871dbdd12c17c became follower at term 1
	2015/10/14 08:29:10 etcdserver: added local member 6a5871dbdd12c17c [http://localhost:2380 http://localhost:7001] to cluster f68652439e3f8f2a
	2015/10/14 08:29:12 raft: 6a5871dbdd12c17c is starting a new election at term 1
	2015/10/14 08:29:12 raft: 6a5871dbdd12c17c became candidate at term 2
	2015/10/14 08:29:12 raft: 6a5871dbdd12c17c received vote from 6a5871dbdd12c17c at term 2
	2015/10/14 08:29:12 raft: 6a5871dbdd12c17c became leader at term 2
	2015/10/14 08:29:12 raft.node: 6a5871dbdd12c17c elected leader 6a5871dbdd12c17c at term 2
	2015/10/14 08:29:12 etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:2379 http://127.0.0.1:4001]} to cluster f68652439e3f8f2a

my docker images are :
* skydns：2015-03-11-001
* etcd：2.0.9 
* kube2sky：1.11

Do you encounter this problem?

**Comment by GuoHui89757** [comment]
@liqinghua188 
E1014 08:46:53.648723       1 reflector.go:136] Failed to list *api.Service: Get http://10.10.103.250:8080/api/v1/services: dial tcp 10.10.103.250:8080: connection refused
E1014 08:46:53.668124       1 reflector.go:136] Failed to list *api.Endpoints: Get http://10.10.103.250:8080/api/v1/endpoints: dial tcp 10.10.103.250:8080: connection refused

Please check kube-apiserver is running correctly.

**Comment by liqinghua188** [comment]
@GuoHui89757 

yes , there are some errors in kube-apiServer logs :

	E1014 16:38:31.706002    7456 reflector.go:183] watch of *api.ResourceQuota ended with: EOF
	E1014 16:38:31.710670    7456 reflector.go:183] watch of *api.Secret ended with: EOF
	E1014 16:38:31.712201    7456 reflector.go:183] watch of *api.LimitRange ended with: EOF
	E1014 16:38:31.713032    7456 reflector.go:183] watch of *api.ServiceAccount ended with: EOF
	E1014 16:38:31.713418    7456 reflector.go:183] watch of *api.Namespace ended with: EOF
	I1014 16:38:31.923474    8519 plugins.go:69] No cloud provider specified.
	I1014 16:38:32.418922    8519 master.go:295] Will report 10.0.3.15 as public IP address.
	E1014 16:38:32.441300    8519 reflector.go:136] Failed to list *api.ResourceQuota: Get http://0.0.0.0:8080/api/v1/resourcequotas: dial tcp 0.0.0.0:8080: connection refused
	E1014 16:38:32.441519    8519 reflector.go:136] Failed to list *api.Secret: Get http://0.0.0.0:8080/api/v1/secrets?fieldSelector=type%3Dkubernetes.io%2Fservice-account-token: dial tcp 0.0.0.0:8080: connection refused
	[restful] 2015/10/14 16:38:32 log.go:30: [restful/swagger] listing is available at https://10.0.3.15:6443/swaggerapi/
	[restful] 2015/10/14 16:38:32 log.go:30: [restful/swagger] https://10.0.3.15:6443/swaggerui/ is mapped to folder /swagger-ui/
	E1014 16:38:32.441820    8519 reflector.go:136] Failed to list *api.ServiceAccount: Get http://0.0.0.0:8080/api/v1/serviceaccounts: dial tcp 0.0.0.0:8080: connection refused
	E1014 16:38:32.441860    8519 reflector.go:136] Failed to list *api.LimitRange: Get http://0.0.0.0:8080/api/v1/limitranges: dial tcp 0.0.0.0:8080: connection refused
	E1014 16:38:32.441906    8519 reflector.go:136] Failed to list *api.Namespace: Get http://0.0.0.0:8080/api/v1/namespaces: dial tcp 0.0.0.0:8080: connection refused
	I1014 16:38:33.249245    8519 server.go:441] Serving securely on 0.0.0.0:6443
	I1014 16:38:33.257864    8519 server.go:483] Serving insecurely on 0.0.0.0:8080

Do you have this problem? I found [issue#11222](https://github.com/kubernetes/kubernetes/issues/11222) are similar .

**Comment by tleyden** [comment]
This was something screwy in my environment.  Closing this ticket for the same reason as mentioned in https://github.com/kubernetes/kubernetes/issues/16244#issuecomment-151594571

@liqinghua188 if you are still seeing reliably reproducible problems, I'd encourage you to open a fresh ticket and provide exact steps to repro.
